### PR TITLE
Improve distributed tracing

### PR DESCRIPTION
**Pull request by allenlsy**
**Description**:

There are some distributed tracing issues with the current demo application.

1. We want to see the tracing graph that connects bookbuyer -> bookstore -> bookwarehouse all together, while currently the chain is broken down into spans.
2. Some users mentioned that the trace ID should be 16 bytes, according to the OpenTracing specification [here](https://www.w3.org/TR/trace-context/#trace-id), while in OSM, the trace ID is 8 bytes, and the first span has the same span ID and trace ID. This can be a potential issue for some customers.

This PR adds extra Zipkin tracing headers to the request in the demo application, so that Envoy can correctly generate tracing information for Jaeger to collect.

**Testing done**:

Deployed the demo application with `demo/run-demo-osm.sh`, and checked tracing graph on Jaeger again.

Screen Shot 2022-01-14 at 2 51 50 PM

Note how the requests are linked together with the same trace ID `337a603dfe4a22ee` in the logs from bookbuyer, bookstore and bookwarehouse.

Screen Shot 2022-01-14 at 2 52 56 PM

Screen Shot 2022-01-14 at 2 53 43 PM

Screen Shot 2022-01-14 at 2 53 59 PM

Related issue: #4430 

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| CI System                  | [ ] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Demo                       | [X] |
| Documentation              | [ ] |
| Egress                     | [ ] |
| Ingress                    | [ ] |
| Install                    | [ ] |
| Networking                 | [ ] |
| Observability              | [X] |
| Performance                | [ ] |
| SMI Policy                 | [ ] |
| Security                   | [ ] |
| Sidecar Injection          | [ ] |
| Tests                      | [ ] |
| Upgrade                    | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

No

2. Is this a breaking change?

No

### CODE REVIEW

File: demo/cmd/bookwarehouse/bookwarehouse.go [hunk 1]
```diff
@@ -49,10 +49,21 @@ func getBooksStockedRecord() Record {
 	return record
 }
 
-// restockBooks decreases the balance of the given bookwarehouse account.
-func restockBooks(w http.ResponseWriter, r *http.Request) {
+func setHeaders(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set(common.IdentityHeader, getIdentity())
 
+	if r != nil {
```
Comment (trstringer) [inline demo/cmd/bookwarehouse/bookwarehouse.go:55]: You can remove a nested layer:

```suggestion
	if r == nil {
	    return
	}
```

**Review by trstringer — CHANGES_REQUESTED** [review]

**Review by shashankram — COMMENTED** [review]
Where is the header being set to 16 bytes instead of 8 as indicated in the commit?

**Comment by allenlsy** [comment]
@shashankram I should update the PR description. That's not addressed in this PR.

**Review by shashankram — APPROVED** [review]
Thanks @allenlsy. Could you also document the required tracing headers in https://main--osm-docs.netlify.app/docs/guides/observability/tracing/

**Review by jaellio — APPROVED** [review]

**Review by trstringer — APPROVED** [review]

**Review by snehachhabria — APPROVED** [review]

**Pull request merged**